Refactor Tradfri switch device

ggravlingen opened the pull request:
## Description:
As I was adding support for the Tradfri blinds, I realized there is a lot of repeated code in the different files of the Tradfri lib. This is the first step in refactoring the code.

**Related issue (if applicable):** relates to #26863

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

ggravlingen committed:
Refactor Tradfri switch device

ggravlingen committed:
Lint

ggravlingen committed:
Lint

ggravlingen commented on homeassistant/components/tradfri/switch.py:
```diff
@@ -1,16 +1,11 @@
-_LOGGER = logging.getLogger(__name__)
 
-TRADFRI_SWITCH_MANAGER = "Tradfri Switch Manager"
+TRADFRI_device_MANAGER = "Tradfri Switch Manager"
```
I don't understand what this is for. Does this have any functionality in Home Assistant's inner workings?

MartinHjelmare replied:
No, it doesn't seem so. We can remove it.

ggravlingen replied:
OK, I removed it

ggravlingen committed:
Removed unused constant

ggravlingen committed:
Add base_class

ggravlingen committed:
Lint

ggravlingen commented:
Pylint throws errors for `no-member` on files I haven't altered. How should I best proceed from here?

MartinHjelmare commented on homeassistant/components/tradfri/base_class.py:
```diff
@@ -0,0 +1,94 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class TradfriBaseDevice:
```
Inherit from `Entity` since we overwrite entity methods.

MartinHjelmare commented on homeassistant/components/tradfri/base_class.py:
```diff
@@ -0,0 +1,94 @@
+        """Start observation of device."""
+        if exc:
+            self._available = False
+            self.async_schedule_update_ha_state()  # pylint: disable=no-member
```
Remove pylint disable.

MartinHjelmare commented on homeassistant/components/tradfri/base_class.py:
```diff
@@ -0,0 +1,94 @@
+                err_callback=self._async_start_observe,
+                duration=0,
+            )
+            self.hass.async_create_task(self._api(cmd))  # pylint: disable=no-member
```
Remove pylint disable.

MartinHjelmare commented on homeassistant/components/tradfri/base_class.py:
```diff
@@ -0,0 +1,94 @@
+    @property
+    def unique_id(self):
+        """Return unique ID for device."""
+        return self._unique_id  # pylint: disable=no-member
```
We need to initialize `self._unique_id` in init method since we use the attribute  here. Remove pylint disable.

MartinHjelmare commented on homeassistant/components/tradfri/base_class.py:
```diff
@@ -0,0 +1,94 @@
+    def _observe_update(self, device):
+        """Receive new state data for this device."""
+        self._refresh(device)
+        self.async_schedule_update_ha_state()  # pylint: disable=no-member
```
Remove pylint disable.

ggravlingen committed:
Improvements after review

ggravlingen commented:
According to Discord chats, there is some error causing pylint to fail. I’ll wait for the fix and rebase.

ggravlingen committed:
Typo

balloob commented on homeassistant/components/tradfri/base_class.py:
```diff
@@ -0,0 +1,96 @@
+            self._available = False
+            self.async_schedule_update_ha_state()
+            _LOGGER.warning("Observation failed for %s", self._name, exc_info=exc)
+
```
Make sure to return or you will still execute the other code ?

```suggestion
           return
```

balloob replied:
And if observation did fail, should you retry right away? Should we have some back off.

ggravlingen replied:
I'm afraid I'm not very familiar with how this part of the code works (looks like you and Lwis wrote it a long time ago.) I know as much that the `_async_start_observe` looks more or less the same in all of the Tradfri device types, so we should be able to move it into the new base class.

However, I wouldn't know what I'm doing when making changes to `_async_start_observe`. It would be a lot of guessing on my end. I'd be happy to just put in whatever you suggests. But that will require time for you to figure out the code. Perhaps a better idea to leave this method as is in the switch-file?

balloob commented on homeassistant/components/tradfri/base_class.py:
```diff
@@ -0,0 +1,96 @@
+                duration=0,
+            )
+            self.hass.async_create_task(self._api(cmd))
+        except PytradfriError as err:
```
What can raise here ? You are using create task so you're not waiting for the result.

ggravlingen commented:
@MartinHjelmare @balloob can this be merged on the basis of my comment above? The intent of this PR, along with a few subsequent ones, is just to remove code that is repeated across the Tradfri-files as well as stale constants.

Once that is done, it will be much less work addressing the issues @balloob raises above.

MartinHjelmare commented:
I think we can merge. @balloob ?
